### PR TITLE
Adding polygon click color toggle example [KOTLIN]

### LINKED ISSUE #924: Add example of selecting and deselecting multiple FillLayer polygons
Would be good to have an example of selecting and deselecting various polygon areas based on `onMapClick()` location. Various countries could be the polygons.

### PR DESCRIPTION
**Pull request by langsmith**
Resolves #924 by adding an example of selecting and deselecting various `FillLayer` polygons. (de)selection is visually represented by fun and random colors.

cc @malwoodsantoro because we'd discussed an example like this a while ago 😄 

cc @jozefgunazer because you commented on #924 👍 

ezgif com-resize (1)

### CODE REVIEW

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    private val BASE_NEIGHBORHOOD_FILL_LAYER_ID = "BASE_NEIGHBORHOOD_FILL_LAYER_ID"
+    private val COLORED_FILL_LAYER_ID = "COLORED_FILL_LAYER_ID"
+    private val LINE_LAYER_ID = "LINE_LAYER_ID"
+    private val NEIGHBORHOOD_POLYGON_SOURCE_ID = "NEIGHBORHOOD_POLYGON_SOURCE_ID"
+    private val PROPERTY_SELECTED = "selected"
+    private val PROPERTY_FILL_COLOR = "fill_color"
+    private val NEIGHBORHOOD_NAME_PROPERTY = "neighborhood_name"
+    private var mapboxMap: MapboxMap? = null
+    private var featureCollection: FeatureCollection? = null
+    private var geoJsonSource: GeoJsonSource? = null
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: most members of this class are constants, either make them actual constant / companion object.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+class PolygonSelectToggleActivity : AppCompatActivity(), MapboxMap.OnMapClickListener {
+
+    private val BASE_NEIGHBORHOOD_FILL_LAYER_ID = "BASE_NEIGHBORHOOD_FILL_LAYER_ID"
+    private val COLORED_FILL_LAYER_ID = "COLORED_FILL_LAYER_ID"
+    private val LINE_LAYER_ID = "LINE_LAYER_ID"
+    private val NEIGHBORHOOD_POLYGON_SOURCE_ID = "NEIGHBORHOOD_POLYGON_SOURCE_ID"
+    private val PROPERTY_SELECTED = "selected"
+    private val PROPERTY_FILL_COLOR = "fill_color"
+    private val NEIGHBORHOOD_NAME_PROPERTY = "neighborhood_name"
+    private var mapboxMap: MapboxMap? = null
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: If you are sure code execution will only occur after mapboxMap is initialized, you can make this lateInit.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                LoadGeoJsonDataTask(this).execute()
+                mapboxMap.addOnMapClickListener(this)
+                Toast.makeText(this, getString(R.string.tap_on_neighborhood),
+                        Toast.LENGTH_SHORT).show()
+            }
+        }
+    }
+
+    override fun onMapClick(point: LatLng): Boolean {
+        return handleClickIcon(mapboxMap?.projection!!.toScreenLocation(point))
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: projection is annotated with `@NonNull` no need for `!!`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /**
+     * Sets up all of the sources and layers needed for this example
+     *
+     * @param collection the FeatureCollection returned by the data loading AsyncTask.
+     */
+    fun setUpData(collection: FeatureCollection) {
+        if (mapboxMap == null) {
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: mapboxMap will allways be non null

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /**
+     * This method handles click events for SymbolLayer symbols.
+     * When a SymbolLayer icon is clicked, we adjust that feature to the selected state.
+     * @param screenPoint the point on screen clicked
+     */
+    private fun handleClickIcon(screenPoint: PointF): Boolean {
+        val features = mapboxMap?.queryRenderedFeatures(screenPoint, BASE_NEIGHBORHOOD_FILL_LAYER_ID)
+        if (features?.isEmpty() == false) {
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: == false is not needed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+            val featureList = featureCollection?.features()
+            for (i in featureList!!.indices) {
+                if (featureList[i].getStringProperty(NEIGHBORHOOD_NAME_PROPERTY) == name) {
+                    if (featureSelectStatus(i)) {
+                        setFeatureSelectState(featureList.get(i), false)
+                    } else {
+                        setSelected(i)
+                    }
+                }
+            }
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: there might be a cleaner Koltin way of handling this iteration as a for i loop

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        return if (featureCollection == null) {
+            false
+        } else featureCollection?.features()!![index].getBooleanProperty(PROPERTY_SELECTED)
+    }
+
+    /**
+     * Updates the display of data on the map after the FeatureCollection has been modified
+     */
+    private fun refreshSource() {
+        if (geoJsonSource != null && featureCollection != null) {
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: null checking is not needed as in next line you are using `?`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+                    // Load GeoJSON file
+                    val `is` = context.assets.open(filename)
+                    val size = `is`.available()
+                    val buffer = ByteArray(size)
+                    `is`.read(buffer)
+                    `is`.close()
+                    return String(buffer)
+                } catch (exception: Exception) {
+                    throw RuntimeException(exception)
+                }
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: Reading files from assets in kotlin can be made simpler:
For example:
``` 
val file_name = "qjsonfile.json"
val json_string = application.assets.open(file_name).bufferedReader().use{
            it.readText()
        }
```

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 2 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 3 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 4 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 5 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 7 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 1 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: Fixed.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 8 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: Cool. Adjusted to:

```kotlin
companion object {
            internal fun loadGeoJsonFromAsset(context: Context, filename: String): String {
                try {
                    // Load GeoJSON file
                    context.assets.open(filename).bufferedReader().use{
                        return it.readText()
                    }
                } catch (exception: Exception) {
                    throw RuntimeException(exception)
                }
            }
        }
```

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt (hunk 6 above)
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/PolygonSelectToggleActivity.kt]: replaced with

```kotlin
            (0 until featureList!!.size).iterator().forEach {
                if (featureList[it].getStringProperty(NEIGHBORHOOD_NAME_PROPERTY) == name) {
                    if (featureSelectStatus(it)) {
                        setFeatureSelectState(featureList.get(it), false)
                    } else {
                        setSelected(it)
                    }
                }
            }
```

**Comment by langsmith** [comment]
Ok @tobrun all feedback has been addressed

**Pull request merged**